backend/xivdpscalc/pps/sample_sch_rotation.py

```diff
@@ -0,0 +1,71 @@
+        """
+        if self.index < len(self.rotation):
+            self.index += 1
+            return self.rotation[self.index - 1]
+        return self.default_action
+
+class SampleSchRotation(SchRotation): #pylint: disable=too-few-public-methods
+    """
+    An example rotation exibiting more dynamic decision rotation generation.
+    This is not meant to be a perfect replication of a good sch rotation, this is primarily for testing and demonstration.
+    """
+
+    def __init__(self):
+        self.rotation = [SchAction.BIOLYSIS, SchAction.AETHERFLOW, SchAction.ENERGYDRAIN,
+                         SchAction.BROIL3, SchAction.BROIL3, SchAction.RUIN2, SchAction.CHAINSTRATAGEM, SchAction.ENERGYDRAIN,
+                         SchAction.BROIL3, SchAction.RUIN2, SchAction.ENERGYDRAIN, SchAction.SWIFTCAST, SchAction.BROIL3,
+                         SchAction.DISSIPATION, SchAction.ENERGYDRAIN, SchAction.BROIL3, SchAction.RUIN2, SchAction.ENERGYDRAIN]
+        self.index = 0
+
+    def get_action(self, current_time: ElapsedTime, cooldowns: dict[SchAction, float],
+                   remaining_effect_durations: dict[SchEffect, float], resources: dict[SchResource, int]) -> SchAction:
+        """
+        Returns a SchAction based on an arbitrary set of rules.
+        """
+        if self.index < len(self.rotation):
+            self.index += 1
+            return self.rotation[self.index - 1]
+        elif cooldowns[SchAction.BROIL3] >= 0.75:
+            # this is a weaving window
+            # if we can avoid clipping, use ED
+            if resources[SchResource.AETHERFLOW] > 0 and cooldowns[SchAction.ENERGYDRAIN] + 0.75 <= cooldowns[SchAction.BROIL3]:
+                return SchAction.ENERGYDRAIN
+            elif cooldowns[SchAction.CHAINSTRATAGEM] + 0.75 <= cooldowns[SchAction.BROIL3]:
+                return SchAction.CHAINSTRATAGEM
+            elif resources[SchResource.AETHERFLOW] <= 0 and cooldowns[SchAction.DISSIPATION] + 0.75 <= cooldowns[SchAction.BROIL3]:
+                return SchAction.DISSIPATION
+            elif resources[SchResource.AETHERFLOW] <= 0 and cooldowns[SchAction.AETHERFLOW] + 0.75 <= cooldowns[SchAction.BROIL3]:
+                return SchAction.AETHERFLOW
+            elif cooldowns[SchAction.SWIFTCAST] + 0.75 <= cooldowns[SchAction.BROIL3]:
+                return SchAction.SWIFTCAST
+        elif remaining_effect_durations[SchEffect.BIOLYSIS] <= 0 or cooldowns[SchAction.AETHERFLOW] <= 0.75:
+            return SchAction.BIOLYSIS
+        elif cooldowns[SchAction.CHAINSTRATAGEM] < 1.5 or \
+            ((cooldowns[SchAction.DISSIPATION] < 1.5) and resources[SchResource.AETHERFLOW] <= 1):
+            return SchAction.RUIN2
+        elif (cooldowns[SchAction.AETHERFLOW] < 30 or cooldowns[SchAction.DISSIPATION] < 30) and \
+            resources[SchResource.AETHERFLOW] > 1 and \
+            cooldowns[SchAction.ENERGYDRAIN] <= 0:
+            return SchAction.RUIN2
+
+        return SchAction.BROIL3
```

Comment:
This is actually a lot of different returns. If you wanted to keep the `if` / `elif` items, you could do a

```python
next_action: SchAction = SchAction.BROIL3
```

at the top and only reassign it you need to; the final return could be at the bottom of the function then.

It might be worth talking through the logic to see if we can simplify this a bit.

Reply:
I added some comments to outline the line of reasoning a bit.  I'm not sure how easy this is to reduce to a simpler form, but if you have suggestions please feel free to toss them my way